File: morris/ai.py
from __future__ import annotations

import math
import random
from dataclasses import dataclass
from typing import Callable, Optional, Tuple

from .game_state import GameState, Move
from .constants import PLAYER_WHITE, PLAYER_BLACK


@dataclass
class AIDifficulty:
	name: str
	depth: int
	use_alpha_beta: bool
	allow_random: bool = False


EASY = AIDifficulty("Easy", depth=1, use_alpha_beta=False, allow_random=True)
MEDIUM = AIDifficulty("Medium", depth=3, use_alpha_beta=True)
HARD = AIDifficulty("Hard", depth=5, use_alpha_beta=True)


def evaluate(state: GameState, maximizing_player: int) -> float:
	# Basic heuristic combining: pieces, mobility, mills potential
	opponent = -maximizing_player
	my_pieces = state.num_pieces(maximizing_player)
	opp_pieces = state.num_pieces(opponent)
	if state.winner == maximizing_player:
		return 10_000.0
	if state.winner == opponent:
		return -10_000.0
	# Mobility (approximate): number of legal moves when it is player's turn if we simulate
	# We compute mobility by temporarily toggling turn
	current_to_move = state.to_move
	state.to_move = maximizing_player
	my_moves = len(state.legal_moves())
	state.to_move = opponent
	opp_moves = len(state.legal_moves())
	state.to_move = current_to_move
	value = 0.0
	value += 100.0 * (my_pieces - opp_pieces)
	value += 2.0 * (my_moves - opp_moves)
	# Prefer exiting placing sooner (material on board)
	value += 1.0 * (9 - (state.white_in_hand if maximizing_player == PLAYER_WHITE else state.black_in_hand))
	return value


def minimax(state: GameState, depth: int, alpha: float, beta: float, maximizing_player: int, use_alpha_beta: bool) -> Tuple[float, Optional[Move]]:
	if depth == 0 or state.winner is not None:
		return evaluate(state, maximizing_player), None
	moves = state.legal_moves()
	if not moves:
		# No legal moves -> gameover for current player
		clone = state.clone()
		clone.winner = -clone.to_move
		return evaluate(clone, maximizing_player), None
	best_move: Optional[Move] = None
	if state.to_move == maximizing_player:
		best_val = -math.inf
		for mv in moves:
			child = state.clone()
			child.apply_move(mv)
			val, _ = minimax(child, depth - 1, alpha, beta, maximizing_player, use_alpha_beta)
			if val > best_val:
				best_val = val
				best_move = mv
			if use_alpha_beta:
				alpha = max(alpha, best_val)
				if beta <= alpha:
					break
		return best_val, best_move
	else:
		best_val = math.inf
		for mv in moves:
			child = state.clone()
			child.apply_move(mv)
			val, _ = minimax(child, depth - 1, alpha, beta, maximizing_player, use_alpha_beta)
			if val < best_val:
				best_val = val
				best_move = mv
			if use_alpha_beta:
				beta = min(beta, best_val)
				if beta <= alpha:
					break
		return best_val, best_move


def choose_move(state: GameState, difficulty: AIDifficulty) -> Optional[Move]:
	moves = state.legal_moves()
	if not moves:
		return None
	if difficulty.allow_random:
		# 50% random on easy for variety
		if random.random() < 0.5:
			return random.choice(moves)
	value, move = minimax(state, depth=difficulty.depth, alpha=-math.inf, beta=math.inf, maximizing_player=state.to_move, use_alpha_beta=difficulty.use_alpha_beta)
	return move if move is not None else random.choice(moves)


def hint(state: GameState, difficulty: AIDifficulty = MEDIUM) -> Optional[Move]:
	return choose_move(state, difficulty)

File: web/js/ai.js
import { PLAYER_WHITE, PLAYER_BLACK } from './constants.js';
import { Move } from './state.js';

export class AIDifficulty{ constructor(name, depth, useAlphaBeta, allowRandom=false){ this.name=name; this.depth=depth; this.useAlphaBeta=useAlphaBeta; this.allowRandom=allowRandom; } }
export const EASY = new AIDifficulty('Easy', 1, false, true);
export const MEDIUM = new AIDifficulty('Medium', 3, true, false);
export const HARD = new AIDifficulty('Hard', 5, true, false);

export function evaluate(state, maximizingPlayer){
  const opponent = -maximizingPlayer;
  const myPieces = state.numPieces(maximizingPlayer);
  const oppPieces = state.numPieces(opponent);
  if(state.winner === maximizingPlayer) return 10000.0;
  if(state.winner === opponent) return -10000.0;
  const currentToMove = state.toMove;
  state.toMove = maximizingPlayer; const myMoves = state.legalMoves().length;
  state.toMove = opponent; const oppMoves = state.legalMoves().length;
  state.toMove = currentToMove;
  let value = 0.0;
  value += 100.0 * (myPieces - oppPieces);
  value += 2.0 * (myMoves - oppMoves);
  value += 1.0 * (9 - (maximizingPlayer === PLAYER_WHITE ? state.whiteInHand : state.blackInHand));
  return value;
}

export function minimax(state, depth, alpha, beta, maximizingPlayer, useAlphaBeta){
  if(depth===0 || state.winner!==null){
    return [evaluate(state, maximizingPlayer), null];
  }
  const moves = state.legalMoves();
  if(moves.length===0){
    const clone = state.clone();
    clone.winner = -clone.toMove;
    return [evaluate(clone, maximizingPlayer), null];
  }
  let bestMove = null;
  if(state.toMove === maximizingPlayer){
    let bestVal = -Infinity;
    for(const mv of moves){
      const child = state.clone();
      child.applyMove(mv);
      const [val] = minimax(child, depth-1, alpha, beta, maximizingPlayer, useAlphaBeta);
      if(val > bestVal){ bestVal = val; bestMove = mv; }
      if(useAlphaBeta){ alpha = Math.max(alpha, bestVal); if(beta <= alpha) break; }
    }
    return [bestVal, bestMove];
  } else {
    let bestVal = Infinity;
    for(const mv of moves){
      const child = state.clone();
      child.applyMove(mv);
      const [val] = minimax(child, depth-1, alpha, beta, maximizingPlayer, useAlphaBeta);
      if(val < bestVal){ bestVal = val; bestMove = mv; }
      if(useAlphaBeta){ beta = Math.min(beta, bestVal); if(beta <= alpha) break; }
    }
    return [bestVal, bestMove];
  }
}

export function chooseMove(state, difficulty){
  const moves = state.legalMoves();
  if(moves.length===0) return null;
  if(difficulty.allowRandom && Math.random()<0.5) return moves[Math.floor(Math.random()*moves.length)];
  const [value, move] = minimax(state, difficulty.depth, -Infinity, Infinity, state.toMove, difficulty.useAlphaBeta);
  return move ?? moves[Math.floor(Math.random()*moves.length)];
}

export function hint(state, difficulty=MEDIUM){
  return chooseMove(state, difficulty);
}
